### ISSUE TITLE
fix: GoogLeNet and MobileNetV1 fieldwise init non-copyable struct error in train.mojo

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#4519]
## Problem

Two example training files fail to compile with --Werror due to non-copyable struct in fieldwise init:

### examples/googlenet-cifar10/train.mojo
```
examples/googlenet-cifar10/train.mojo:89:8: error: cannot synthesize fieldwise init because field 'field0' has non-copyable and non-movable type 'GoogLeNet'
    fn compute_batch_loss(
       ^
```

### examples/mobilenetv1-cifar10/train.mojo
```
examples/mobilenetv1-cifar10/train.mojo:89:8: error: cannot synthesize fieldwise init because field 'field0' has non-copyable and non-movable type 'MobileNetV1'
    fn compute_batch_loss(
       ^
```

## Root Cause

The `GoogLeNet` and `MobileNetV1` structs are non-copyable/non-movable, but a struct containing them
as a field is using `@fieldwise_init` (or implicit fieldwise init synthesis), which requires all
fields to be copyable.

## Fix Options

1. Add explicit `__init__` that takes ownership (using `^`) of the model field
2. Change the containing struct to use `var model: UnsafePointer[GoogLeNet]` if copying isn't needed
3. Mark the containing struct with `@register_passable` if appropriate

## Affected Files

- `examples/googlenet-cifar10/train.mojo`
- `examples/mobilenetv1-cifar10/train.mojo`

## Context

Found during comprehensive `--Werror` compilation audit (PR #4512, branch `batch/low-complexity-fixes`).
Related issues: #4513 (ExTensor circular type), #4514 (linker -lm), #4515 (FP8/BF8 API).

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have everything I need. The VGG16 approach — inline batch loop without a closure — is the correct pattern. Here's the implementation plan:

---

# Implementation Plan: Issue #4519 — Fix Non-Copyable Struct Fieldwise Init Error

## 1. Objective

Fix compile errors in `examples/googlenet-cifar10/train.mojo` and `examples/mobilenetv1-cifar10/train.mojo` where Mojo's closure synthesis fails because `GoogLeNet` and `MobileNetV1` are non-copyable/non-movable structs that cannot be captured by a nested `fn compute_batch_loss(...)` closure.

## 2. Root Cause Analysis

When Mojo encounters a nested `fn` inside a function, it synthesizes a hidden closure struct whose fields hold captured variables. Both files define:

```mojo
fn compute_batch_loss(
    batch_images: ExTensor, batch_labels: ExTensor
) raises -> Float32:
    var logits = model.forward(batch_images, training=True)
    ...
```

This nested function captures `model` from the outer scope. Mojo synthesizes a fieldwise init for this closure struct with `field0: GoogLeNet` (or `MobileNetV1`). Since `GoogLeNet`/`MobileNetV1` are non-copyable and non-movable, the fieldwise init synthesis fails.

**Error (line 89/90 = the nested `fn` definition):**
```
error: cannot synthesize fieldwise init because field 'field0' has non-copyable and non-movable type 'GoogLeNet'
```

## 3. Approach

**Replace the nested closure + `run_epoch_manual` pattern with an inline batch loop**, matching the pattern used in `examples/vgg16-cifar10/train.mojo`.

The inline loop directly calls the batch body without capturing anything in a closure struct, eliminating the fieldwise init synthesis entirely.

The `run_epoch_manual` accepts a `fn (ExTensor, ExTensor) raises -> Float32` function pointer — but Mojo's nested `fn` with a capture requires a fieldwise init of a closure aggregate. Since the model is non-copyable, we must avoid closures entirely.

## 4. Files to Create

None.

## 5. Files to Modify

### `examples/googlenet-cifar10/train.mojo`

**Change**: Replace the `var loop = TrainingLoop(...)` + `fn compute_batch_loss(...)` + `loop.run_epoch_manual(...)` pattern in `train_epoch` with an inline batch loop.

**Specific change** (lines ~287–318):

```mojo
# BEFORE:
    var loop = TrainingLoop(log_interval=100)

    fn compute_batch_loss(
        batch_images: ExTensor, batch_labels: ExTensor
    ) raises -> Float32:
        var logits = model.forward(batch_images, training=True)
        var loss = cross_entropy(logits, batch_labels)
        var loss_value = loss._data.bitcast[Float32]()[0]
        # ...
        return loss_value

    var avg_loss = loop.run_epoch_manual(
        train_images,
        train_labels,
        batch_size=batch_size,
        compute_batch_loss=compute_batch_loss,
        epoch=epoch + 1,
        total_epochs=200,
    )

    return avg_loss
```

```mojo
# AFTER:
    var num_samples = train_images.shape()[0]
    var num_batches = compute_num_batches(num_samples, batch_size)
    var total_loss = Float32(0.0)
    var log_interval = 100

    print("Epoch [", epoch + 1, "/", 200, "]")

    for batch_idx in range(num_batches):
        var start_idx = batch_idx * batch_size
        var batch_pair = extract_batch_pair(
            train_images, train_labels, start_idx, batch_size
        )
        var batch_images = batch_pair[0]
        var batch_labels = batch_pair[1]

        var logits = model.forward(batch_images, training=True)
        var loss = cross_entropy(logits, batch_labels)
        var loss_value = loss._data.bitcast[Float32]()[0]
        # Backward/update (stub, same as before)
        total_loss += loss_value

        if (batch_idx + 1) % log_interval == 0:
            var avg = total_loss / Float32(batch_idx + 1)
            print("  Batch [", batch_idx + 1, "/", num_batches, "] - Loss: ", avg)

    var avg_loss = total_loss / Float32(num_batches)
    print("  Average Loss: ", avg_loss)
    return avg_loss
```

Also remove the now-unused `from shared.training.loops import TrainingLoop` import (and `TrainingLoop` if not used elsewhere in the file).

### `examples/mobilenetv1-cifar10/train.mojo`

**Same fix** as above, applied to the `train_epoch` function (lines ~85–118).

Replace:
```mojo
    var loop = TrainingLoop(log_interval=100)
    fn compute_batch_loss(...) raises -> Float32:
        ...
    var avg_loss = loop.run_epoch_manual(..., compute_batch_loss=compute_batch_loss, ...)
    return avg_loss
```

With an inline batch loop identical in structure to the googlenet fix above, using `model.forward(batch_images, training=True)` and `cross_entropy(logits, batch_labels)`.

Remove unused `from shared.training.loops import TrainingLoop` import.

## 6. Implementation Order

1. **Fix `examples/mobilenetv1-cifar10/train.mojo`** — simpler file (294 lines), same closure pattern at lines 86–118. Replace nested fn + `run_epoch_manual` with inline batch loop.

2. **Fix `examples/googlenet-cifar10/train.mojo`** — same pattern at lines 287–319. Replace nested fn + `run_epoch_manual` with inline batch loop.

3. **Remove unused `TrainingLoop` import** from both files (if `TrainingLoop` is no longer referenced anywhere else in each file — verify with grep first).

4. **Run pre-commit hooks** to verify formatting: `just pre-commit`.

5. **Attempt compilation** in CI (or Docker container): `mojo build examples/googlenet-cifar10/train.mojo` and `mojo build examples/mobilenetv1-cifar10/train.mojo` with `--Werror`.

## 7. Verification

- **Compile check**: Both files must compile without error under `--Werror`
- **No behavior change**: The batch loop body is identical in logic — forward pass, cross-entropy, stub backward — just not wrapped in a closure
- **Grep check**: Confirm no remaining `compute_batch_loss` nested fn or `TrainingLoop` import in patched files
- **CI**: Push to branch `3774-conv2d-gradient-checking` (current) or a new branch and confirm the `--Werror` build step passes

## 8. Skills Used

- **`docs-search`** / direct file reads — to understand the `TrainingLoop.run_epoch_manual` signature and identify what the closure captures
- **`validate-mojo-patterns`** (referenced) — non-copyable struct ownership rules; nested `fn` closure capture constraints
- **`mojo-lint-syntax`** (referenced) — confirms that nested `fn` with captured non-copyable fields is invalid in Mojo v0.26.1
- No external skills were invoked during planning; analysis was done via direct file reads and grep

---
*Generated by Claude Code Planner*
